core/typedefs.h

```diff
@@ -126,14 +126,38 @@ T *_nullptr() {
 #define MIN(m_a, m_b) (((m_a) < (m_b)) ? (m_a) : (m_b))
 #endif
 
+#ifndef MIN2
+#define MIN2(m_a, m_b) (Vector2(MIN(m_a.x, m_b.x), MIN(m_a.y, m_b.y)))
+#endif
+
+#ifndef MIN3
+#define MIN3(m_a, m_b) (Vector3(MIN(m_a.x, m_b.x), MIN(m_a.y, m_b.y), MIN(m_a.z, m_b.z)))
+#endif
+
 #ifndef MAX
 #define MAX(m_a, m_b) (((m_a) > (m_b)) ? (m_a) : (m_b))
 #endif
 
+#ifndef MAX2
+#define MAX2(m_a, m_b) (Vector2(MAX(m_a.x, m_b.x), MAX(m_a.y, m_b.y)))
+#endif
+
+#ifndef MAX3
+#define MAX3(m_a, m_b) (Vector3(MAX(m_a.x, m_b.x), MAX(m_a.y, m_b.y), MAX(m_a.z, m_b.z)))
+#endif
+
 #ifndef CLAMP
 #define CLAMP(m_a, m_min, m_max) (((m_a) < (m_min)) ? (m_min) : (((m_a) > (m_max)) ? m_max : m_a))
 #endif
 
+#ifndef CLAMP2
+#define CLAMP2(m_a, m_min, m_max) (Vector2(CLAMP(m_a.x, m_min.x, m_max.x), CLAMP(m_a.y, m_min.y, m_max.y)))
+#endif
+
+#ifndef CLAMP3
+#define CLAMP3(m_a, m_min, m_max) (Vector3(CLAMP(m_a.x, m_min.x, m_max.x), CLAMP(m_a.y, m_min.y, m_max.y), CLAMP(m_a.z, m_min.z, m_max.z)))
+#endif
+
```

Comment:
Why are we not using templates instead?

Reply:
Because it leads to ambiguity problems in a lot of places.

Reply:
I can try with inline overloading however..

Reply:
Do you have an example I can use to understand this?

Reply:
Well actually Vector2 and Vector3 are not defined in this file - so we need to push the include to this file (which may be undesired), I've tried to use templates and there is a result:
So I've tried to push this code

```
template <class T>
inline T MIN(T a, T b) {
	return (a < b) ? a : b;
}

template <class T>
inline T MIN2(const T &a, const T &b) {
	return T(MIN(a.x, b.x), MIN(a.y, b.y));
}

template <class T>
inline T MIN3(const T &a, const T &b) {
	return T(MIN(a.x, b.x), MIN(a.y, b.y), MIN(a.z, b.z));
}

template <class T>
inline T MAX(T a, T b) {
	return (a > b) ? a : b;
}

template <class T>
inline T MAX2(const T &a, const T &b) {
	return T(MAX(a.x, b.x), MAX(a.y, b.y));
}

template <class T>
inline T MAX3(const T &a, const T &b) {
	return T(MAX(a.x, b.x), MAX(a.y, b.y), MAX(a.z, b.z));
}

template <class T>
inline T CLAMP(T value, T min, T max) {
	return (value < min) ? min : ((value > max) ? max : value);
}

template <class T>
inline T CLAMP2(const T &value, const T &min, const T &max) {
	return T(clamp(value.x, min.x, max.x), clamp(value.y, min.y, max.y));
}

template <class T>
inline T CLAMP3(const T &value, const T &min, const T &max) {
	return T(clamp(value.x, min.x, max.x), clamp(value.y, min.y, max.y), clamp(value.z, min.z, max.z));
}
```
to typedefs.h

![image](https://user-images.githubusercontent.com/3036176/63094879-48332880-bf72-11e9-98ef-3a88268b4fcf.png)

Sorry for non-english output..